plugins/auth-backend/src/providers/oauth2/provider.ts

```diff
@@ -130,46 +144,124 @@ export class OAuth2AuthProvider implements OAuthHandlers {
       refreshToken: updatedRefreshToken,
     } = refreshTokenResponse;
 
-    const rawProfile = await executeFetchUserProfileStrategy(
+    const fullProfile = await executeFetchUserProfileStrategy(
       this._strategy,
       accessToken,
     );
-    const profile = makeProfileInfo(rawProfile, params.id_token);
 
-    return this.populateIdentity({
-      providerInfo: {
-        accessToken,
-        refreshToken: updatedRefreshToken,
-        idToken: params.id_token,
-        expiresInSeconds: params.expires_in,
-        scope: params.scope,
-      },
-      profile,
+    return this.handleResult({
+      fullProfile,
+      params,
+      accessToken,
+      refreshToken: updatedRefreshToken,
     });
   }
 
-  // Use this function to grab the user profile info from the token
-  // Then populate the profile with it
-  private async populateIdentity(
-    response: OAuthResponse,
-  ): Promise<OAuthResponse> {
-    const { profile } = response;
+  private async handleResult(result: OAuthResult) {
+    const { profile } = await this.authHandler(result);
+
+    const response: OAuthResponse = {
+      providerInfo: {
+        idToken: result.params.id_token,
+        accessToken: result.accessToken,
+        scope: result.params.scope,
+        expiresInSeconds: result.params.expires_in,
+      },
+      profile,
+    };
 
-    if (!profile.email) {
-      throw new Error('Profile does not contain an email');
+    if (this.signInResolver) {
+      response.backstageIdentity = await this.signInResolver(
+        {
+          result,
+          profile,
+        },
+        {
+          tokenIssuer: this.tokenIssuer,
+          catalogIdentityClient: this.catalogIdentityClient,
+          logger: this.logger,
+        },
+      );
     }
-    const id = profile.email.split('@')[0];
 
-    return { ...response, backstageIdentity: { id } };
+    return response;
   }
 }
 
-export type OAuth2ProviderOptions = {};
+export const oAuth2EmailSignInResolver: SignInResolver<OAuthResult> = async (
+  info,
+  ctx,
+) => {
+  const { profile } = info;
+
+  if (!profile.email) {
+    throw new Error('OAuth2 profile contained no email');
+  }
+
+  const entity = await ctx.catalogIdentityClient.findUser({
+    annotations: {
+      'oauth2/email': profile.email,
+    },
+  });
+
+  const claims = getEntityClaims(entity);
+  const token = await ctx.tokenIssuer.issueToken({ claims });
+
+  return { id: entity.metadata.name, entity, token };
+};
+
+export const oAuth2DefaultSignInResolver: SignInResolver<OAuthResult> = async (
+  info,
+  ctx,
+) => {
+  const { profile } = info;
+
+  if (!profile.email) {
+    throw new Error('Profile contained no email');
+  }
+
+  let userId: string;
+
+  try {
+    const entity = await ctx.catalogIdentityClient.findUser({
+      annotations: {
+        'oauth2/email': profile.email,
```

Comment:
I'm still unsure about this prefix. In the Google and Microsoft provider they used google.com and microsoft.com as prefix. Any suggestions or is this oke?

Reply:
To be honest I'm leaning towards leaving it out. That or possibly something like `backstage.io/email`? but I think there's a pretty big risk that that ends up being confusing.

On that topic the default resolver should implement the existing logic only, meaning just the email split. It's so we have backwards compatibility but it will eventually be phased out.

Reply:
Hi @Rugvip,

First off, thank you for your response and time to look into it. Due to vacations a late response from myself. Sorry for that. I will implement the default resolver behavior. I will look into it, and update the PR.
  
Will update soon. 

Best regards,
Pepijn